Add Versioning tutorial for Cellar with AWS CLI and MinIO (#516)
CleverCloud/documentation: content/doc/addons/cellar.md

You may use SSG (Static Site Generated) to  dynamically generate your content during your build.

## Versioning

### What is bucket versioning

> Versioning allow you to store multiple variants of an object inside of your bucket. When you upload a file with the same name as one already present in the bucket, Cellar  store a version of the previous content. This protect you from unintended action such as the accidental deletion of a file. 

Reviewer: ```suggestion
Versioning enables you to store multiple versions of an object in your bucket. When you upload a file with the same name as an existing one, Cellar automatically saves the previous version of that object. This safeguards your data against accidental actions, such as deleting a file by mistake.
```

> When you store a new version of a file, only the most recent version is visible. The previous one are hidden but still stored inside of the bucket. This allow you to recover an older version of a file in case of error.

Reviewer: ```suggestion
Only the latest version of a file is visible, while previous versions remain hidden but securely stored in the bucket. This ensures you can easily restore an earlier version if needed.
```

### How does versioning work

> Versioning can't be enabled for file, it can only be enabled for an entire bucket.  

Reviewer: ```suggestion
Versioning can't be enabled for a single file: it will cover the entire bucket.
```

A bucket can be in one of three state :

 - Unversioned (The default)
 - Versioning-enabled
 - Versioning-suspended

> You enable and suspend versioning at the bucket level. After you version-enable a bucket, it can never return to an unversioned state. But you can suspend versioning on that bucket.

Reviewer: ```suggestion
You enable and suspend versioning at the bucket level. Once you enable versioning on a bucket, it can never return to an unversioned state. But you can suspend versioning on that bucket at any time. 
```

> Once versioning is enabled, any object you add have a unique version ID. Object that already existed before enabling versioning have a version ID of `null`.

Reviewer: ```suggestion
Once versioning is enabled, any object you add have a unique version ID. Object that were already existing before enabling versioning have a version ID of `null`.

For that reason, we do recommend you to enable versioning when creating a new bucket. It will be easier to navigate through version IDs.
```

> {{< callout type="warning">}}
> Versioning can quickly take up a lot of space since multiple version of an object are stored in the bucket.  
> {{< /callout>}}

Reviewer: ```suggestion
> [!WARNING]
> Versioning can quickly take up a lot of space since multiple version of an object are stored in the bucket.  
```

{{< tabs items="MinIO,  AWS CLI" >}}

  {{< tab >}}
  
  To use [minIO](https://min.io/docs/minio/linux/reference/minio-mc.html#command-mc), you must first creare an alias.

  ```sh
  mc alias set <ALIAS_NAME> https://cellar-c2.services.clever-cloud.com <ACCESS_KEY> <SECRET_KEY>
  ```

  ### Activate Versioning with MinIO

>   To activate Versioning, you can use MinIO and the following command.

Reviewer 2: ```suggestion
  To activate versioning, you can use MinIO and the following command.
```

  ```sh
  mc version enable <alias>/<bucket-name>
  ```

  If you want to suspend versioning, you can replace `enable` by `suspend`

  You can check that versioning is enabled for your bucket with :

  ```sh
  mc version info <alias>/<bucket-name> --json
  ```

  ### How to use versioning

>   When versioning is enabled, the object added automatically have a Unique identifier added to them. Only the latest version of an object is shown with a `mc ls <alias>/<bucket-name>`.

Reviewer: ```suggestion
When versioning is enabled, the newly added object is automatically provided with a unique identifier. Only the latest version of an object is shown with a `mc ls <alias>/<bucket-name>`.
```

>   #### List versioned Object

Reviewer: ```suggestion
  #### List versioned objects
```

  If you need to list all the object in your bucket, including the different versions of the files stored in it, you can run : 

  ```sh
  mc ls --versions --recursive <alias>/<bucket-name>
  ```

  You can list all the version of a specific file with 

  ```sh
  mc ls --versions <alias>/<bucket-name>/<object_name>
  ```

  #### Get a version of an object 

>   You can get the specific version of an object using it's version ID obtained with the previous command

Reviewer: ```suggestion
  You can get the specific version of an object using its version ID obtained using the previous command.
```

  ```sh
  mc get --vid <version_id> <alias>/<bucket-name>/<object_name> <path/to/save/file>
  ```

  #### Delete a version of an object 

>   You can delete the specific version of an object using it's version ID

Reviewer: ```suggestion
  You can delete the specific version of an object using its version ID.
```

  ```sh
  mc rm --vid <version_id> <alias>/<bucket-name>/<object_name>
  ```

  {{< /tab >}}

  {{< tab >}}

>   The following command assume you have configured your AWS CLI and added an alias as showed earlier in the section `With AWS CLI `.

Reviewer: ```suggestion
  The following command assumes you have configured your AWS CLI and added an alias as shown earlier in the section [Creating a bucket with AWS CLI](https://www.clever-cloud.com/developers/doc/addons/cellar/#with-aws-cli)
```

  ### Activate versioning with AWS CLI

  To activate versioning, you can use AWS CLI. You can use the following command to enable it on a bucket.

  ```sh
  aws s3api put-bucket-versioning --bucket <bucket_name> --versioning-configuration Status=Enabled
  ```

>   If you want to turn off versioning, you just need to set `--versioning-configuration` with `Status=Suspended`

Reviewer: ```suggestion
  If you want to turn off versioning, you can use the following:
  
```sh
  aws s3api put-bucket-versioning --bucket <bucket_name> --versioning-configuration Status=Suspended
```
```

>   You can check that versioning is enabled for your bucket with :

Reviewer: ```suggestion
  You can check if versioning is enabled on your bucket with :
```

  ```sh
  aws s3api get-bucket-versioning --bucket <bucket_name>
  ```

  ### How to use versioning

>   When versioning is enabled, the object added automatically have a versionID added to them. Only the latest version of an object if shown.

Reviewer: ```suggestion
  When versioning is enabled, the added object automatically gets a versionID. Only the latest version of an object if shown.
```

  #### List all versioned Object

>   If you need to list all the object in your bucket, including the different versions of the files stored in it, you can run : 

Reviewer: ```suggestion
  If you need to list all the objects in your bucket, including the different versions of the files stored in it, you can use : 
```

  ```sh
  aws s3api list-object-versions --bucket <bucket_name>
  ```

  #### List all the version of a specific Object

>   If you want to find the versions of only a specific object, you can use the following command. In this example, we find the versions of test.txt :

Reviewer: ```suggestion
  If you want to check the different versions of a specific object, you can use the following command. In this example, we find the versions of test.txt :
```

  ```sh
  aws s3api list-object-versions --bucket <bucket_name> --prefix <file_name>
  ```

  #### Get the version of an Object

>   If you want to get the previous version of an object, you need the versionID that you can get with the two previous command. You can then use :

Reviewer: ```suggestion
  If you want to retrieve the previous version of an object, you need the versionID that you can get with any of the two previous command. You can use :
```
